presc/report/runner.py

```diff
@@ -75,30 +54,40 @@ def _updated_jb_toc(report_config):
     Returns the updated JB TOC file as a YAML-formatted string that can be
     written to a _toc.yml.
     """
-    # Rather than parsing as YAML and dealing with the nested structure,
-    # just look through the raw lines of the file for report page entries
-    # starting with "- file: "
     with open(REPORT_SOURCE_PATH / JB_TOC_FILENAME) as f:
-        toc_lines = f.readlines()
+        toc_str = f.read()
 
-    stripped_lines = [(i, x.strip()) for i, x in enumerate(toc_lines)]
-    all_pages = {i: x[8:] for i, x in stripped_lines if x.startswith("- file: ")}
+    stripped_lines = [x.strip() for x in toc_str.split("\n")]
+    all_pages = [x[8:] for x in stripped_lines if x.startswith("- file: ")]
     incl_pages = include_exclude_list(
-        list(all_pages.values()),
-        report_config["evaluations_include"],
-        report_config["evaluations_exclude"],
+        all_pages,
+        report_config["evaluations_include"].get(),
+        report_config["evaluations_exclude"].get(),
     )
-    # Landing page should always be included.
     if "landing" not in incl_pages:
         incl_pages.append("landing")
 
-    # Retain all TOC lines except those corresponding to excluded pages.
-    filtered_toc = [
-        x
-        for i, x in enumerate(toc_lines)
-        if i not in all_pages or all_pages[i] in incl_pages
-    ]
-    return "".join(filtered_toc)
+    def drop_file_entries(toc_part, incl_list):
+        if isinstance(toc_part, list):
+            filtered = [drop_file_entries(x, incl_list) for x in toc_part]
+            return [x for x in filtered if x]
+        if isinstance(toc_part, dict):
+            if "file" in toc_part.keys() and toc_part["file"] not in incl_list:
+                return None
+            else:
+                filtered = {
+                    k: drop_file_entries(v, incl_list) for k, v in toc_part.items()
+                }
+                return {k: v for k, v in filtered.items() if v}
+        else:
+            return toc_part
+
+    toc = yaml.load(toc_str, Loader=yaml.FullLoader)
+    # Drop excluded file entries from the TOC and any subsections that are left empty.
+    filtered_toc = drop_file_entries(toc, incl_pages)
+    # Drop any top-level entries that don't have files in their hierarchy
+    filtered_toc = [x for x in filtered_toc if "file" in str(x)]
+    return yaml.dump(filtered_toc)
```

Comment:
This seems overly complicated.  There is not way to to apply the include and exclude lists directly to the toc?